### DISCUSSION TITLE
calling another function

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#92] [Q&A]
I try to set up the word functions as bifs and call it from another rxas file. I looked up the testsubstr.rxas function calling substr.rxbin and understand the logic:
         ```
          load r3,4  /* there are four arguments*/
          load r4,"abc"
          load r5,2 /* second argument, startpos */
          load r6,6 /* third argument, length */
          load r7,"." /* fourth argument, pad */ 
          call r1,substr(),r3
          say r1
```
call r1,substr(),r3  the r3 parm identifies the number of parameters, which consecutively use the registers after r3: r4,45,r6,r7. 

My problem is:

```
/* words function */
       .globals=0
main() .locals=8 .expose=test.main
    load r3,1  /* there are four arguments*/
    load r4,"The quick brown fox jumps over the lazy dog"
    call r1,words(),r3
    say r1
```
seems ok, but I got this asm error:
```
C:\Users\PeterJ\CLionProjects\CREXX_F030\cmake-build-debug\assembler\rxas.exe testword

Error Severity 1
6:12 - Error at "words()", unknown procedure

Process finished with exit code 0
```
What else is necessary to define an external function?

**Comment by rvjansen** [comment]
It needs to be global.words instead of testl.main

**Reply by Peter-Jacob** [reply]
I have that, in words.rxas:
```
          .globals=0
words()  .locals=8 .expose=global.words
    /* string    = arg1    */
...
```

the problem is in asm of testword.rxas, where it say words unknown

**Comment by rvjansen** [comment]
you also need a words() .expose-global.words in testwords

**Comment by Peter-Jacob** [comment]
I don't see the error!

C:\Users\PeterJ\CLionProjects\CREXX_F030\cmake-build-debug\interpreter\rxvm.exe testword
Unimplemented symbol: global.words

Process finished with exit code -1

testword.rxas
```
/* Test word function */
        .globals=0
words() .expose=global.words
main()  .locals=8
    load r3,1  /* there are four arguments*/
    load r4,"The quick brown fox jumps over the lazy dog"
    call r1,words(),r3
    say r1
	
words.rxas	
/*
     Words function
 */
         .globals=0
words()  .locals=8 .expose=global.words
    /* string    = arg1    */

   copy  r1,a1
   load  r3,0   /* offset in string */
   load  r4,43  /* length of string */
   load  r6,0   /* word count       */
loop:
   fndnblnk r3,r1,r3   /* find next non blank offset */
   ilt r5,r3,0         /* if -1, nothing found, end search */
   brt break,r5
   inc r6              /* else increase word count  */
                       /* offset of word is in R3   */
   fndblnk r3,r1,r3    /* from offset find next blank offset */
   ilt r5,r3,0         /* if -1, nothing found, end search   */
   brt break,r5
   bct loop,r4,r3      /* then look for next non blank char  */
break:
   load r7,"Words found "
   say r7
   itos r6
   say r6
   ret r6
```

**Comment by rvjansen** [comment]
You did not push fndnblnk - but when I comment that out, it works.

**Comment by Peter-Jacob** [comment]
I pushed the changes I have

**Comment by rvjansen** [comment]
➜  rxfns git:(feature/f0030) ✗ rxvm testwords words
Words found
9
9
[1]    7226 segmentation fault  rxvm testwords words
➜  rxfns git:(feature/f0030) ✗

**Comment by rvjansen** [comment]
Ah wait, you work from test

**Comment by rvjansen** [comment]
But same result ;-)

**Comment by Peter-Jacob** [comment]
ok, time to dump the pc out of the window!

**Comment by Peter-Jacob** [comment]
storing/running all rxas and rxbin functions out of the f0030/.../tests directory is ok, isn't it?

**Comment by rvjansen** [comment]
Yes, should be. As long as the executables are in your path, or the whole path is specified, you should be ok

**Comment by adesutherland** [comment]
Can you try

    rxvm testwords words

You have to list each rxbin file ... let me know if it helps!

**Reply by Peter-Jacob** [reply]
```
C:\Users\PeterJ\CLionProjects\CREXX_F030\cmake-build-debug\interpreter\rxvm.exe testword words


Process finished with exit code 0

```
works fine
